File: src/ObjectManager/Object.Tes/TesAssetManager.cs
using OA.Core;
using OA.Tes.FilePacks;
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace OA.Tes
{
    public enum GameId
    {
        // tes
        Morrowind = 1,
        Oblivion,
        Skyrim,
        SkyrimSE,
        SkyrimVR,
        // fallout
        Fallout3,
        FalloutNV,
        Fallout4,
        Fallout4VR,
    }

    public class TesAssetManager : IAssetManager
    {
        public Task<IAssetPack> GetAssetPack(Uri uri)
        {
            switch (uri.Scheme)
            {
                case "game":
                    {
                        var localPath = uri.LocalPath.Substring(1);
                        var gameId = StringToGameId(uri.Host);
                        var filePaths = FileManager.GetFilePaths(localPath, gameId);
                        if (filePaths == null)
                            throw new InvalidOperationException($"{gameId} not available");
                        var pack = (IAssetPack)new TesAssetPack(filePaths, null);
                        return Task.FromResult(pack);
                    }
                case "file":
                    {
                        var localPath = uri.LocalPath;
                        var gameId = StringToGameId(uri.Host);
                        var filePaths = FileManager.GetFilePaths(localPath, gameId);
                        var pack = (IAssetPack)new TesAssetPack(filePaths, null);
                        return Task.FromResult(pack);
                    }
                default:
                    {
                        var pack = (IAssetPack)new TesAssetPack((string[])null, null);
                        return Task.FromResult(pack);
                    }
            }
        }

        public Task<IDataPack> GetDataPack(Uri uri)
        {
            switch (uri.Scheme)
            {
                case "game":
                    {
                        var localPath = uri.LocalPath.Substring(1);
                        var gameId = StringToGameId(uri.Host);
                        var filePath = FileManager.GetFilePath(localPath, gameId);
                        var pack = (IDataPack)new TesDataPack(filePath, null, gameId);
                        return Task.FromResult(pack);
                    }
                case "file":
                    {
                        var localPath = uri.LocalPath;
                        if (!File.Exists(localPath))
                            throw new IndexOutOfRangeException("file not found");
                        var gameId = StringToGameId(uri.Fragment);
                        var pack = (IDataPack)new TesDataPack(localPath, null, gameId);
                        return Task.FromResult(pack);
                    }
                default:
                    {
                        var gameId = StringToGameId(uri.Fragment);
                        var pack = (IDataPack)new TesDataPack(null, null, gameId);
                        return Task.FromResult(pack);
                    }
            }
        }

        public ICellManager GetCellManager(IAssetPack asset, IDataPack data, TemporalLoadBalancer loadBalancer)
        {
            return new TesCellManager((TesAssetPack)asset, (TesDataPack)data, loadBalancer);
        }

        static GameId StringToGameId(string key)
        {
            if (key.StartsWith("#")) key = key.Substring(1);
            var game = Enum.GetNames(typeof(GameId)).FirstOrDefault(x => string.Compare(x, key, true) == 0);
            if (game == null)
                throw new ArgumentOutOfRangeException("key", key);
            return (GameId)Enum.Parse(typeof(GameId), game);
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/TesAssetManager.swift
//
//  TesAssetManager.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation

public enum GameId: Int, CustomStringConvertible {
    // tes
    case morrowind = 0
    case oblivion
    case skyrim, skyrimSE, skyrimVR
    // fallout
    case fallout3
    case falloutNV
    case fallout4, fallout4VR

    public var description: String {
        switch self {
        case .morrowind: return "morrowind"
        case .oblivion: return "oblivion"
        case .skyrim: return "skyrim"
        case .skyrimSE: return "skyrimSE"
        case .skyrimVR: return "skyrimVR"
        case .fallout3: return "fallout3"
        case .falloutNV: return "falloutNV"
        case .fallout4: return "fallout4"
        case .fallout4VR: return "fallout4VR"
        default: return "?"
        }
    }
    
    init?(_ value: String) {
        switch value.lowercased() {
        case "morrowind": self = .morrowind
        case "oblivion": self = .oblivion
        case "skyrim": self = .skyrim
        case "skyrimse": self = .skyrimSE
        case "skyrimvr": self = .skyrimVR
        case "fallout3": self = .fallout3
        case "falloutnv": self = .falloutNV
        case "fallout4": self = .fallout4
        case "fallout4vr": self = .fallout4VR
        default: return nil
        }
    }
}

public class TesAssetManager: IAssetManager {    
    public func getAssetPack(_ url: URL?) -> IAssetPack? {
        guard let url = url, let scheme = url.scheme else {
            fatalError("should not happen")
        }
        debugPrint("\(url)")
        switch scheme {
        case "game":
            let localPath = String(url.path[url.path.index(after: url.path.startIndex)..<url.path.endIndex])
            let game = TesAssetManager.stringToGameId(url.host)
            let filePath = FileManager.default.getFilePath(localPath, for: game)
            guard filePath != nil else {
                fatalError("\(game) not available")
            }
            let pack = TesAssetPack(filePath!) as IAssetPack
            return pack
        default: return nil
        }
    }
    
    public func getDataPack(_ url: URL?) -> IDataPack? {
        guard let url = url, let scheme = url.scheme else {
            fatalError("should not happen")
        }
        debugPrint("\(url)")
        switch scheme {
        case "game":
            let localPath = String(url.path[url.path.index(after: url.path.startIndex)..<url.path.endIndex])
            let game = TesAssetManager.stringToGameId(url.host)
            let filePath = FileManager.default.getFilePath(localPath, for: game)
            let pack = TesDataPack(filePath, for: game) as IDataPack
            return pack
        default: return nil
        }
    }
    
    public func getCellManager(asset: IAssetPack, data: IDataPack, loadBalancer: TemporalLoadBalancer) -> ICellManager? {
        return TesCellManager(asset: asset as! TesAssetPack, data: data as! TesDataPack, loadBalancer: loadBalancer)
    }

    static func stringToGameId(_ key: String?) -> GameId {
        guard let key = key else {
            return GameId.morrowind
        }
        var newKey = key
        if newKey.starts(with: "#") {
            newKey.remove(at: newKey.startIndex)
        }
        return GameId(newKey)
    }
}
